File: 0451_sort_characters_by_frequency/frequency_sort.go
package main

/**
 * Time: O(n)
 * Space: O(n)
 */
func frequencySort(s string) string {
	counter := map[rune]int{}
	for _, char := range s {
		counter[char]++
	}
	buckets := map[int][]rune{}
	maxFreq := 0
	for k, v := range counter {
		buckets[v] = append(buckets[v], k)
		if v > maxFreq {
			maxFreq = v
		}
	}
	var res []rune
	for ; maxFreq > 0; maxFreq-- {
		for _, v := range buckets[maxFreq] {
			for i := 0; i < maxFreq; i++ {
				res = append(res, v)
			}
		}
	}
	return string(res)
}

File: 0451_sort_characters_by_frequency/frequency_sort.py
from collections import Counter


def frequency_sort(s: str) -> str:
    """
    Time: O(n)
    Space: O(n)
    """
    return ("".join(char * times for char, times in Counter(s).most_common()))


def frequency_sort_bucket(s: str) -> str:
    """
    Time: O(n)
    Space: O(n)
    """
    buckets, res = [[]]*len(s), ""
    counter = Counter(s)
    for ch, cnt in counter.items():
        buckets[cnt-1] = buckets[cnt-1]+[ch]
    for i in range(len(buckets)-1, -1, -1):
        if buckets[i]:
            res += "".join([c*(i+1) for c in buckets[i]])
    return res
